scripts/migrate_maas_instance_ids.sh

```diff
@@ -0,0 +1,126 @@
+#!/bin/bash
+
+set -e
+
+#This script script is intended to run only against controller machines that
+#have upgraded from a MAAS 1.9 provider to a MAAS 2.x provider. Do not use this
+#script otherwise. The javascript below is horrible. It is designed to run
+#operations in bulk targeting a minimum version of mongo 3.2 which has little
+#functionality in the way of aggregation.
+
+usage() { echo "usage: $0 [-rh]" 1>&2; exit 1; }
+
+while getopts 'rh' flag; do
+    case "${flag}" in
+        r) rollback_flag=true
+           ;;
+        h) usage
+           ;;
+        *) usage
+           ;;
+    esac
+done
+
+if [ $rollback_flag ]
+then
+    #This script warps a MAAS 1.9 style URL around the instance_id. You can run
+    #this script with the -r flag for convenience but you should really just
+    #reload the backup that you have created with mongoexport or otherwise if
+    #you run into trouble.
+    mongo_script=$(cat <<'JAVASCRIPT'
+db.instanceData.aggregate([{
+    "$project": {
+        instanceid: {
+            "$cond": {
+                if: {
+                    "$ne": [{
+                        "$substr": ["$instanceid", 0, 20]
+                    }, "/MAAS/api/1.0/nodes/"]
+                }
+                , then: {
+                    "$concat": ["/MAAS/api/1.0/nodes/", "$instanceid", "/"]
+                }
+                , else: "$instanceid"
+            }
+        }
+        , document: "$$ROOT"
+    }
+}, {
+    "$out": "tmp"
+}]);
+ops = db.instanceData.find().map(function (item) {
```

Comment:
I'm hoping that you tested this live, but I thought juju-mongodb3.2 and juju-mongodb both had the Javascript V8 engine turned off, which means that we can't define new functions to pass into map().
I also thought that we might not have the $substr operator (I thought I tried that in other code).

But maybe the "mongo" client is interpreting the javascript, rather than passing it into the v8 engine in mongod?

As long as you've tested it running inside a machine that you "juju bootstrap" its probably fine. I would just caution that "apt install mongodb-server" gives you a different engine than "juju bootstrap" would.

Reply:
Sure